File: client/src/utils/modules/AutoRoleModule.ts
import { FoxyClient } from "../../structures/types/foxy";

export default class AutoRoleModule {
    public bot: FoxyClient;
    constructor(bot) {
        this.bot = bot;
    }

    async check(member) {
        const guildId = member.guildId;
        const guildInfo = await this.bot.database.getGuild(guildId);
        if (!this.bot.hasGuildPermission(this.bot, guildId, ["MANAGE_ROLES"] || ["ADMINISTRATOR"])) return;
        if (guildInfo.AutoRoleModule.isEnabled) {
            if (member.user.toggles.bot) return;

            const roles = guildInfo.AutoRoleModule.roles;
            if (roles.length > 0) {
                for (const role of roles) {
                    try {
                        setTimeout(async () => {
                            await this.bot.helpers.addRole(guildId, member.id, role, "Auto Role");
                        }, 2000);
                    } catch (error) { }
                }
            }
        }
    }
}

File: foxy/src/main/kotlin/net/cakeyfox/foxy/modules/autorole/AutoRoleModule.kt
package net.cakeyfox.foxy.modules.autorole

import kotlinx.coroutines.delay
import kotlinx.coroutines.suspendCancellableCoroutine
import net.cakeyfox.foxy.FoxyInstance
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.Role
import net.dv8tion.jda.api.events.guild.member.GuildMemberJoinEvent
import net.dv8tion.jda.api.Permission
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

class AutoRoleModule(
    val foxy: FoxyInstance
) {
    suspend fun handleUser(event: GuildMemberJoinEvent) {
        if (!event.guild.selfMember.hasPermission(Permission.MANAGE_ROLES)) return

        val guildData = foxy.database.guild.getGuild(event.guild.id)
        val member = event.member

        if (guildData.AutoRoleModule != null) {
            if (guildData.AutoRoleModule!!.isEnabled) {
                val roleIds = guildData.AutoRoleModule!!.roles

                val rolesToGive = roleIds.mapNotNull { event.guild.getRoleById(it) }
                    .filter { canAssignRole(event.guild, it) }

                rolesToGive.take(5).forEach { role ->
                    assignRoleAsync(event.guild, member, role)
                    delay(200)
                }
            }
        }
    }

    private fun canAssignRole(guild: Guild, role: Role): Boolean {
        val topBotRole = guild.selfMember.roles.maxByOrNull { it.position }?.position ?: 0
        return topBotRole > role.position
    }

    private suspend fun assignRoleAsync(
        guild: Guild,
        member: Member,
        role: Role
    ) {
        suspendCancellableCoroutine { continuation ->
            guild.addRoleToMember(member, role).queue(
                { continuation.resume(Unit) },
                { error -> continuation.resumeWithException(error) }
            )
        }
    }
}
